dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -618,18 +665,89 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
     {
         BOOL foregroundHandled = true;
         THROW_IF_FAILED(InvokeAll(payloadLength, payload, &foregroundHandled));
+        THROW_HR_IF(E_UNEXPECTED, !foregroundHandled);
 
-        if (!foregroundHandled)
+        return S_OK;
+    }
+    CATCH_RETURN();
+
+    void PushNotificationManager::Run(winrt::IBackgroundTaskInstance const& taskInstance)
+    {
+        // This function can be triggered by either OS background infrastructure
+        // or by the PushNotificationsLongRunningProcess.
+        bool firstNotificationReceived{ false };
+        {
+            auto lock{ m_lock.lock_shared() };
+            firstNotificationReceived = m_firstNotificationReceived;
+        }
+        if (!firstNotificationReceived)
```

Comment:
Race condition

Reply:
Should retrieving firstNotificationReceived be lock_exclusive()?

Reply:
What do we do in AppNotificationManager for races?

Reply:
Looks like we lock_exclusive() the entire block.

Reply:
Yeah. You have your answer then :)